File: platform/ios/xcode/FullStacked/InstanceEditor.swift
//
//  InstanceEditor.swift
//  FullStacked
//
//  Created by Charles-Philippe Lepage on 2024-03-22.
//

import Foundation
import SwiftyJSON
import UIKit

let paths = NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true);
let documentsDirectory = paths.first!

class InstanceEditor: Instance {
    static var singleton: InstanceEditor?
    
    init(){
        let editorDirectory = Bundle.main.path(forResource: "build", ofType: nil)!
        super.init(adapter: AdapterEditor(baseDirectory: editorDirectory))
        self.webview.isOpaque = false
        InstanceEditor.singleton = self
    }
}

class AdapterEditor: Adapter {
    let rootDirectory = documentsDirectory
    private let baseJSFile = Bundle.main.path(forResource: "base", ofType: "js", inDirectory: "js")!
    let cacheDirectory = FileManager.default.temporaryDirectory.absoluteString
    let configDirectory = ".config/fullstacked"
    let nodeModulesDirectory: String
    let fsEditor: AdapterFS
    let bonjour = Bonjour()
    let multipeer = Multipeer()
    
    init(baseDirectory: String) {
        self.nodeModulesDirectory = configDirectory + "/node_modules"
        self.fsEditor = AdapterFS(baseDirectory: self.rootDirectory);
        super.init(projectId: nil, baseDirectory: baseDirectory)
        
        
        self.bonjour.onPeerNearby = {eventType, peerNearbyBonjour in
            let message = [
                "eventType": eventType,
                "peerNearby": [
                    "type": 1,
                    "peer": [
                        "id": peerNearbyBonjour.id,
                        "name": peerNearbyBonjour.name
                    ],
                    "addresses": peerNearbyBonjour.addresses,
                    "port": peerNearbyBonjour.port
                ]
            ]
            InstanceEditor.singleton?.push(messageType: "peerNearby", message: JSON(message).rawString()!)
        }
        self.multipeer.onPeerNearby = {eventType, peerNearbyMultipeer in
            let message = [
                "eventType": eventType,
                "peerNearby": [
                    "type": 2,
                    "peer": [
                        "id": peerNearbyMultipeer.peer.id,
                        "name": peerNearbyMultipeer.peer.name
                    ],
                    "id": peerNearbyMultipeer.id
                ]
            ]
            InstanceEditor.singleton?.push(messageType: "peerNearby", message: JSON(message).rawString()!)
        }
        
        self.multipeer.onPeerConnection = { id, type, state in
            let message = [
                "id": id,
                "type": type,
                "state": state
            ]
            InstanceEditor.singleton?.push(messageType: "peerConnection", message: JSON(message).rawString()!)
        }
        
        self.multipeer.onPeerData = { id, data in
            let message = [
                "id": id,
                "data": data
            ]
            InstanceEditor.singleton?.push(messageType: "peerData", message: JSON(message).rawString()!)
        }
    }
    
    override func callAdapterMethod(methodPath: [String.SubSequence], args: [Any?], done: @escaping (_ maybeData: Any?) -> Void) {
        if(methodPath.count == 0) {
            return done(nil)
        }
        
        switch(methodPath.first) {
        case "migrate":
            let project = args[0] as! JSON
            let oldPath = project["location"].stringValue
            let newPath = project["id"].stringValue
            return done(self.fsEditor.rename(oldPath: oldPath, newPath: newPath))
            case "directories":
                switch(methodPath[1]) {
                    case "rootDirectory": return done(self.rootDirectory)
                    case "cacheDirectory": return done(self.cacheDirectory)
                    case "configDirectory": return done(self.configDirectory)
                    case "nodeModulesDirectory": return done(self.nodeModulesDirectory)
                    default: break
                }
                break
            case "fs":
            
            var absolutePath = false
            if(args[0] is String && (args[0] as! String).contains("@types/react")) {
                print(args[0] as! String)
                print("ici")
            }
            if(args.count > 0 && args[0] is JSON && (args[0] as! JSON)["absolutePath"].boolValue){
                absolutePath = true
            } else if(args.count > 1 && args[1] is JSON && (args[1] as! JSON)["absolutePath"].boolValue) {
                absolutePath = true
            } else if (args.count > 2 && args[2] is JSON && (args[2] as! JSON)["absolutePath"].boolValue) {
                absolutePath = true
            }
            
            if (absolutePath) {
                switch(methodPath[1]){
                case "readFile":
                    let utf8 = args.count > 1 && (args[1] as! JSON)["encoding"].stringValue == "utf8"
                    return done(self.fsEditor.readFile(path: args[0] as! String, utf8: utf8))
                case "writeFile":
                    return done(self.fsEditor.writeFile(file: args[0] as! String, strOrData: args[1]!))
                case "writeFileMulti":
                    var i = 1;
                    while(i < args.count) {
                        let file = args[i] as! String
                        let data = args[i + 1]!
                        let maybeError = self.fsEditor.writeFile(file: file, strOrData: data)
                        if(maybeError is AdapterError){
                            return done(maybeError)
                        }
                        i += 2
                    }
                    return done(true)
                case "unlink": return done(self.fsEditor.unlink(path: args[0] as! String))
                case "readdir":
                    var withFileTypes = false
                    var recursive = false
                    if(args.count > 1) {
                        let options = args[1] as! JSON
                        withFileTypes = options["withFileTypes"].boolValue
                        recursive = options["recursive"].boolValue
                    }
                    let files = self.fsEditor.readdir(path: args[0] as! String, withFileTypes: withFileTypes, recursive: recursive)
                    return done(files)
                    case "mkdir": return done(self.fsEditor.mkdir(path: args[0] as! String))
                    case "rmdir": return done(self.fsEditor.rmdir(path: args[0] as! String))
                    case "stat": return done(self.fsEditor.stat(path: args[0] as! String))
                    case "lstat": return done(self.fsEditor.lstat(path: args[0] as! String))
                    case "exists":
                        let exists = self.fsEditor.exists(path: args[0] as! String)
                        return done(exists == nil ? false : exists)
                    case "rename":
                        return done(self.fsEditor.rename(oldPath: args[0] as! String, newPath: args[1] as! String))
                    default: break;
                }
            }
            break
            case "esbuild":
                switch(methodPath[1]) {
                case "version":
                    var versionStrPtr = UnsafeMutablePointer<Int8>(nil)
                    version(&versionStrPtr)
                    let versionStr = String.init(cString: versionStrPtr!, encoding: .utf8)
                    return done(versionStr)
                case "baseJS":
                    let content = FileManager.default.contents(atPath: self.baseJSFile)!
                    return done(String(data: content, encoding: .utf8)!)
                case "check": return done("1")
                case "install": break
                case "tmpFile":
                    switch(methodPath[2]) {
                    case "write":
                        let path = self.cacheDirectory + (args[0] as! String)
                        let data = (args[1] as! String).data(using: .utf8)!
                        try! data.write(to: URL(string: path)!)
                        return done(String(path.dropFirst("file://".count)))
                    case "unlink":
                        let path = self.cacheDirectory + (args[0] as! String)
                        try! FileManager.default.removeItem(at: URL(string: path)!)
                        return done(true)
                    default: break;
                    }
                case "build":
                    let inputPtr = UnsafeMutablePointer<Int8>(mutating: ((args[0] as! String) as NSString).utf8String)
                    let outPtr = UnsafeMutablePointer<Int8>(mutating: ("index" as NSString).utf8String)
                    let outdirPtr = UnsafeMutablePointer<Int8>(mutating: ((args[1] as! String) as NSString).utf8String)
                    let nodePathPtr = UnsafeMutablePointer<Int8>(mutating: (self.rootDirectory + "/" + self.nodeModulesDirectory as NSString).utf8String)
                    
                    var errorsPtr = UnsafeMutablePointer<Int8>(nil)
                    
                    build(inputPtr,
                          outPtr,
                          outdirPtr,
                          nodePathPtr,
                          &errorsPtr)
                                
                    let errors = String.init(cString: errorsPtr!, encoding: .utf8)!
                    if(!errors.isEmpty) {
                        return done(JSON(parseJSON: errors))
                    }
                
                    return done(1)
                default: break
                }
                break
        case "run":
            let project = args[0] as! JSON
            let projectDirectory = self.rootDirectory + "/" + project["location"].stringValue
                
            let runningInstance = RunningInstances.singleton?.getInstance(projectDirectory: projectDirectory)
        
            if(runningInstance != nil) {
                runningInstance!.webview.reload()
            } else {
                let project = Project(
                    location: projectDirectory,
                    id: project["id"].stringValue,
                    title: project["title"].stringValue)
                RunningInstances.singleton?.addInstance(instance: Instance(project: project))
            }
        
            return done(true)
        case "open":
            let project = args[0] as! JSON
            let projectLocation = self.rootDirectory + "/" + project["location"].stringValue
            UIApplication.shared.open(URL(string: "shareddocuments://" + projectLocation)!)
            return done(true)
        case "connectivity":
            switch(methodPath[1]) {
            case "infos":
                return done(false)
            case "name":
                return done(UIDevice.current.name)
            case "peers":
                switch(methodPath[2]){
                case "nearby":
                    var peersNearby: [JSON] = []
                    let peersNearbyBonjour = self.bonjour.getPeersNearby().arrayValue
                    let peersNearbyMultipeer = self.multipeer.getPeersNearby().arrayValue
                    
                    for peerNearby in peersNearbyBonjour {
                        peersNearby.append(peerNearby)
                    }
                    for peerNearby in peersNearbyMultipeer {
                        peersNearby.append(peerNearby)
                    }
                    
                    return done(JSON(peersNearby))
                default: break
                }
            case "advertise":
                switch(methodPath[2]){
                case "start":
                    let peer = args[0] as! JSON
                    self.multipeer.startAdvertising(id: peer["id"].stringValue, name: peer["name"].stringValue)
                    return done(true)
                case "stop":
                    self.multipeer.stopAdvertising()
                    return done(true)
                default: break
                }
            case "browse":
                switch(methodPath[2]){
                case "start":
                    self.bonjour.startBrowsing()
                    self.multipeer.startBrowsing()
                    return done(true)
                case "peerNearbyIsDead":
                    self.bonjour.peerNearbyIsDead(id: args[0] as! String)
                    return done(true)
                case "stop":
                    self.bonjour.stopBrowsing()
                    self.multipeer.stopBrowsing()
                    return done(true)
                default: break
                }
            case "open":
                let me = (args[1] as! JSON)
                self.multipeer.open(id: (args[0] as! String), meId: me["id"].stringValue, meName: me["name"].stringValue)
                return done(true)
            case "disconnect":
                self.multipeer.disconnect(id: args[0] as! String)
                return done(true)
            case "trustConnection":
                self.multipeer.trustConnection(id: args[0] as! String)
                return done(true)
            case "send":
                self.multipeer.send(id: args[0] as! String, data: args[1] as! String, pairing: args[2] as! Bool)
                return done(true)
            case "convey":
                let data = args[1] as! String
                let projectId = args[0] as! String;
                RunningInstances.singleton!.instances.forEach({instance in
                    if(instance.adapter.projectId == projectId) {
                        instance.push(messageType: "peerData", message: data)
                    }
                })
                return done(true)
            default: break
            }
        default: break
        }
        
        return super.callAdapterMethod(methodPath: methodPath, args: args, done: done)
    }
}

File: platform/android/studio/app/src/main/java/org/fullstacked/editor/InstanceEditor.kt
package org.fullstacked.editor

import android.app.DownloadManager
import android.content.Intent
import android.os.Environment
import android.provider.DocumentsContract
import org.fullstacked.editor.connectivity.Bonjour
import org.fullstacked.editor.connectivity.Peer
import org.fullstacked.editor.connectivity.PeerNearby
import org.fullstacked.editor.connectivity.WSS
import org.json.JSONArray
import org.json.JSONObject
import java.io.ByteArrayOutputStream
import java.io.File
import java.io.IOException
import java.io.InputStream
import kotlin.math.abs


val editorProject = Project(
    location = "",
    id = "org.fullstacked.editor",
    title = "FullStacked"
)

class InstanceEditor(var context: MainActivity) : Instance(
    project = editorProject,
    init = false
) {
    companion object {
        lateinit var singleton: InstanceEditor
        fun initialized(): Boolean = this::singleton.isInitialized
    }

    val instances = mutableListOf<Instance>()

    init {
        singleton = this

        this.adapter = AdapterEditor(
            projectId = editorProject.id,
            baseDirectory = this.context.filesDir.toString()
        )

        this.render()
    }

    override fun render() {
        val webView = createWebView(
            ctx = this.context,
            adapter = this.adapter,
            isEditor = true
        )

        if(this.webViewState != null) {
            webView.restoreState(this.webViewState!!)
        }

        this.webViewId = webView.id

        this.context.setContentView(webView)
    }
}

class AdapterEditor(
    projectId: String,
    private val baseDirectory: String,
): Adapter(projectId, baseDirectory) {
    private val webSocketServer = WSS()
    private val bonjour = Bonjour(webSocketServer)

    companion object {
        init {
            System.loadLibrary("editor")
        }
    }

    private external fun version(): String

    private external fun build(
        input: String,
        outdir: String,
        nodePath: String,
    ): String

    override fun getFile(path: String): InputStream? {
        return try {
            InstanceEditor.singleton.context.assets.open(path)
        } catch (e: IOException) {
            null
        }
    }

    private fun directoriesSwitch(directory: String): String? {
        when (directory) {
            "rootDirectory" -> return this.baseDirectory
            "cacheDirectory" -> return InstanceEditor.singleton.context.cacheDir.toString()
            "configDirectory" -> return ".config"
            "nodeModulesDirectory" -> return ".cache/node_modules"
        }

        return null
    }

    private fun esbuildSwitch(methodPath: List<String>, args: List<Any?>) : Any? {
        when (methodPath.first()) {
            "version" -> return version()
            "check" -> return true
            "baseJS" -> return convertInputStreamToString(InstanceEditor.singleton.context.assets.open("base.js"))
            "tmpFile" -> {
                when (methodPath.elementAt(1)) {
                    "write" -> {
                        val filePath = InstanceEditor.singleton.context.cacheDir.toString() + "/" + (args[0] as String)
                        val file = File(filePath)
                        file.writeText(args[1] as String)
                        return filePath
                    }
                    "unlink" -> {
                        val filePath = InstanceEditor.singleton.context.cacheDir.toString() + "/" + (args[0] as String)
                        val file = File(filePath)
                        file.delete()
                        return true
                    }
                }
            }
            "build" -> {
                val errors = build(
                    input = args[0] as String,
                    outdir = args[1] as String,
                    nodePath = this.baseDirectory + "/.cache/node_modules"
                )

                if (errors.isEmpty())
                    return true

                return JSONArray(errors)
            }
        }

        return null
    }

    private fun connectivitySwitch(methodPath: List<String>, args: List<Any?>) : Any? {
        when (methodPath.first()) {
            "name" -> return android.os.Build.MODEL
            "infos" -> {
                val infos = JSONObject()
                infos.put("port", this.webSocketServer.port)

                val addresses = Bonjour.getIpAddress()
                val infoAddresses = JSONArray()
                for (address in addresses)
                    infoAddresses.put(address)

                val networkInterface = JSONObject()
                networkInterface.put("name", "Active Network")
                networkInterface.put("addresses", infoAddresses)

                val networkInterfaces = JSONArray()
                networkInterfaces.put(networkInterface)

                infos.put("networkInterfaces", networkInterfaces)

                return infos
            }
            "peers" -> {
                when (methodPath.elementAt(1)) {
                    "nearby" -> {
                        val json = JSONArray()

                        this.bonjour.getPeersNearby().forEach { peerNearby ->
                            json.put(Bonjour.serializePeerNearby(peerNearby)) }

                        return json
                    }
                }
            }
            "advertise" -> {
                when (methodPath.elementAt(1)) {
                    "start" -> {
                        val peerJSON = args[0] as JSONObject
                        this.bonjour.startAdvertising(Peer(
                            id = peerJSON.getString("id"),
                            name = peerJSON.getString("name")
                        ))
                        return true
                    }
                    "stop" -> {
                        this.bonjour.stopAdvertising()
                        return true
                    }
                }
            }
            "browse" -> {
                when (methodPath.elementAt(1)) {
                    "start" -> {
                        this.bonjour.startBrowsing()
                        return true
                    }
                    "peerNearbyIsDead" -> {
                        this.bonjour.peerNearbyIsDead(args[0] as String)
                        return true
                    }
                    "stop" -> {
                        this.bonjour.stopBrowsing()
                        return true
                    }
                }
            }
            "open" -> {}
            "disconnect" -> {
                this.webSocketServer.disconnect(args[0] as String)
                return true
            }
            "trustConnection" -> {
                this.webSocketServer.trustConnection(args[0] as String)
                return true
            }
            "send" -> {
                this.webSocketServer.send(args[0] as String, args[1] as String, args[2] as Boolean)
                return true
            }
            "convey" -> {
                val projectId = args[0] as String
                val data = args[1] as String

                InstanceEditor.singleton.instances.forEach { instance ->
                    if(instance.project.id == projectId) {
                        instance.push("peerData", data)
                    }
                }

                return true
            }
        }
        return null
    }

    private fun run(project: JSONObject): Boolean {
        InstanceEditor.singleton.context.runOnUiThread {
            val instance = Instance(
                project = Project(
                    id = project.getString("id"),
                    title = project.getString("title"),
                    location = project.getString("location")
                )
            )
            InstanceEditor.singleton.instances.add(instance)
        }

        return true
    }

    override fun callAdapterMethod(methodPath: ArrayList<String>, args: List<Any?>): Any? {
        when (methodPath.first()) {
            "migrate" -> {
                val projectJSON = args[0] as JSONObject
                val oldPath = projectJSON.getString("location")
                val newPath = projectJSON.getString("id")
                return this.fs.rename(oldPath, newPath)
            }
            "directories" -> return this.directoriesSwitch(methodPath.elementAt(1))
            "esbuild" -> return this.esbuildSwitch(methodPath.subList(1, methodPath.size), args)
            "connectivity" -> return this.connectivitySwitch(methodPath.subList(1, methodPath.size), args)
            "run" -> return this.run(args[0] as JSONObject)
            "open" -> {
                val project = args[0] as JSONObject
                val title = project.getString("title")
                val location = project.getString("location")

                val file = File(this.baseDirectory + "/" + location + "/" + title + ".zip")
                if(!file.exists()) return false

                val out = File(Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS).absolutePath + "/" + title + ".zip")
                file.copyTo(out, true)

                val intent = Intent(DownloadManager.ACTION_VIEW_DOWNLOADS)
                InstanceEditor.singleton.context.startActivity(intent)
                return true
            }
            "fs" -> {
                var absolutePath = false
                var utf8 = false

                for(arg in args) {
                    if (arg is JSONObject) {
                        absolutePath = absolutePath || arg.optBoolean("absolutePath")
                        utf8 = utf8 || arg.optString("encoding") == "utf8"
                    }
                }

                if (absolutePath) return super.callAdapterMethod(methodPath, args)

                if (args.isEmpty()) return null

                val file = this.getFile(args[0] as String)

                if (file != null && utf8) {
                    return convertInputStreamToString(file)
                }

                return file
            }
        }


        return super.callAdapterMethod(methodPath, args)
    }
}


fun convertInputStreamToString(inputStream: InputStream): String {
    val result = ByteArrayOutputStream()
    val buffer = ByteArray(DEFAULT_BUFFER_SIZE)
    var length: Int
    while ((inputStream.read(buffer).also { length = it }) != -1) {
        result.write(buffer, 0, length)
    }

    return result.toString("UTF-8")
}